Mission RTL flight mode

Torjusba opened the pull request:
Breaking out Mission RTL in a separate flight mode in order to separate it from rallypoint RTL in failsafe actions, and be able to failsafe from mission RTL to rallypoint RTL.
The main change in behavior from the upstream "mission rtl_type" is the direction selection logic and not respecting jump-to items to avoid an infinite loop bug.

Falls back to RTL if not FW.
Since MissionRTL is not supported in MR mode, it triggers RTL as soon as the aircraft backtransitions. This is fine since the end result is the same (landed at the target location)

Edgecases that are not handled, but considered out of scope:
- If the drone falls back to RTL because it's MR, but we then transition to FW, it wil un-fallback to M-RTL. I don't see this happening in practice, and if it does it's not a big issue. 
- What if the winch waypoint is at the end of a long route transport mission? Then the aircraft will reverse, but not have enough charge to fly back home, and the battery failsafe will trigger.

Torjusba committed:
mission RTL flight mode

Breaking out Mission RTL in a separate flight mode in order to separate
it from rallypoint RTL in failsafe actions, and be able to failsafe from
mission RTL to rallypoint RTL.

Torjusba committed:
convenience script for running SITL in docker

oystub commented on Tools/aviant/sitl_docker.sh:
```diff
@@ -0,0 +1,42 @@
+if [[ $(id -u) != 0 ]]; then
+    echo "SITL must be root to have write permission from inside the docker container"
+    exit 1
+fi
```
~~This is only the case for rootless?~~

I actually don't understand this comment at all... Can't you just check if the user has permission for docker?

I guess the point is, if running in rootless, the uid translation means that the "px4" user inside docker does not have the same permissions as the local user?

oystub replied:
The easiest way of solving this would be to add --user=root to the run command if docker is in rootless.

(And make the image behave ok if run as  `root`, not only `px4`)

With rootless, root inside container atomatically maps to current user outside, so permissions work out really nicely.

oystub commented on src/lib/events/enums.json:
```diff
@@ -417,22 +417,26 @@
                             "description": "Hold"
                         },
                         "6": {
+                            "name": "mission_rtl",
```
I'm a bit confused about the difference between `mission_rtl` and `auto_mission_rtl`. Will remove this comment if I understand

oystub replied:
Also why they have nearly the same description

Torjusba replied:
It is not displayed nicely in the github review view, but they are different enums.

oystub replied:
Aha, lol, right

oystub commented on src/lib/events/enums.json:
```diff
@@ -417,22 +417,26 @@
+                            "name": "mission_rtl",
+                            "description": "MissionRTL"
+                        },
+                        "7": {
```
Do we really want to change all numbers?

Torjusba replied:
We have to, unfortunately. Actions with higher numbers take precedence, see the original definition in `framework.h:52`

oystub commented on src/lib/modes/ui.hpp:
```diff
@@ -62,8 +62,9 @@ static inline uint32_t getValidNavStates()
 	       (1u << vehicle_status_s::NAVIGATION_STATE_AUTO_PRECLAND) |
 	       (1u << vehicle_status_s::NAVIGATION_STATE_ORBIT) |
 	       (1u << vehicle_status_s::NAVIGATION_STATE_AUTO_VTOL_TAKEOFF);
+	(1u << vehicle_status_s::NAVIGATION_STATE_AUTO_MISSION_RTL);
```
Should be ORed with the previous lines

oystub commented on src/modules/commander/ModeUtil/conversions.hpp:
```diff
@@ -93,11 +93,13 @@ static inline navigation_mode_t navigation_mode(uint8_t nav_state)
 }
 
-} // namespace mode_util
+} // namespace mode_util
```
newline

oystub commented on src/modules/commander/ModeUtil/mode_requirements.cpp:
```diff
@@ -113,6 +113,16 @@ void getModeRequirements(uint8_t vehicle_type, failsafe_flags_s &flags)
 	setRequirement(vehicle_status_s::NAVIGATION_STATE_AUTO_RTL, flags.mode_req_home_position);
 	setRequirement(vehicle_status_s::NAVIGATION_STATE_AUTO_RTL, flags.mode_req_prevent_arming);
 
+	// NAVIGATION_STATE_AUTO_MISSION_RTL
+	setRequirement(vehicle_status_s::NAVIGATION_STATE_AUTO_MISSION_RTL, flags.mode_req_angular_velocity);
+	setRequirement(vehicle_status_s::NAVIGATION_STATE_AUTO_MISSION_RTL, flags.mode_req_attitude);
+	setRequirement(vehicle_status_s::NAVIGATION_STATE_AUTO_MISSION_RTL, flags.mode_req_local_position);
+	setRequirement(vehicle_status_s::NAVIGATION_STATE_AUTO_MISSION_RTL, flags.mode_req_global_position);
+	setRequirement(vehicle_status_s::NAVIGATION_STATE_AUTO_MISSION_RTL, flags.mode_req_local_alt);
+	setRequirement(vehicle_status_s::NAVIGATION_STATE_AUTO_MISSION_RTL, flags.mode_req_home_position);
+	setRequirement(vehicle_status_s::NAVIGATION_STATE_AUTO_MISSION_RTL, flags.mode_req_prevent_arming);
+	setRequirement(vehicle_status_s::NAVIGATION_STATE_AUTO_MISSION_RTL, flags.mode_req_mission);
```
Should this have 
```
flags.mode_req_wind_and_flight_time_compliance
```
to better match mission?

oystub commented on src/modules/commander/failsafe/framework.h:
```diff
@@ -103,6 +104,8 @@ class FailsafeBase: public ModuleParams
 
 		case Action::Hold: return "Hold";
 
+		case Action::MissionRTL: return "Mission Return";
```
Do we want to call it "Mission Return" consistently, or match the existing "Return/RTL" (since it is different in different places)? Now we are also inconsistent, but in different places.

Torjusba replied:
I have tried to match the existing "Return/RTL", but failed. Will fix

oystub commented on src/modules/navigator/mission_base.cpp:
```diff
@@ -1412,4 +1412,4 @@ void MissionBase::updateMissionAltAfterHomeChanged()
 		_home_update_counter = _navigator->get_home_position()->update_count;
 	}
-}
+}
```
newline

oystub commented on src/modules/navigator/mission_rtl.cpp:
```diff
@@ -0,0 +1,199 @@
+	// Does not matter if this is Fast or FastReverse
+	// One needs to exist before activation to have a mission
+	// We can read the closest waypoint from
+	_mission_handle = new RtlMissionFast(_navigator);
```
Very useful comment!

oystub commented on src/modules/navigator/mission_rtl.cpp:
```diff
@@ -0,0 +1,199 @@
+
+	float km_backwards{0};
+	float km_forwards{0};
+	bool any_winch_commands_forwards{false};
+	bool any_winch_commands_backwards{false};
```
Could we have an enum with commands to skip?

oystub replied:
(this is a bit unrelated to exactly these lines)

oystub commented on src/modules/navigator/mission_rtl.cpp:
```diff
@@ -0,0 +1,199 @@
+		mission_item_s *previous_item = nullptr;
+
+		// Note: Home location is not part of the mission here
+		// That is fine, just means that the transition distance is not counted
```
I didn't intuitively understand what is meant by "transition distance" here.

Torjusba replied:
I meant "distance from takeoff to the first waypoint". Will fix

oystub commented on src/modules/navigator/mission_rtl.cpp:
```diff
@@ -0,0 +1,199 @@
+			// Assume that all mission items with non-zero lat or lon are valid
+			// Note that this check assumes that we don't fly near 0,0, which is a fair assumption for now.
+			const bool nonzero_lat = (PX4_ISFINITE(current_item->lat) && abs(current_item->lat) >= 0.01);
+			const bool nonzero_lon = (PX4_ISFINITE(current_item->lon) && abs(current_item->lon) >= 0.01);
```
What is the existing PX4 behaviour around zeroes in waypoints? Any specific reason for this guard?

There is't any reason to do >=0.01 here? These are not the result of calculations, we just want to see if they are set or not, so a pure == comparison is safe.

Torjusba replied:
Some commands don't have lat/lon, e.g. MAV_CMD_DO_WINCH. So we have to check this.

I agree with the ==, but -Werror=float-equal won't let me do it. I don't think it hurts to check it this way, it's not very expensive.

oystub commented on src/modules/navigator/mission_rtl.cpp:
```diff
@@ -0,0 +1,199 @@
+				&dist_xy_m,
+				&dist_z_m
+			);
+			const float km = dist_xy_m * 1e-3f;
```
Would like a comment about why dist_z is not used.

oystub replied:
Why do we need to convert to km at all?

I feel like meters is used everywhere else in px4. I think it's cleaner to just keep that and then convert just for the print statements.

Torjusba replied:
mistake from me, started using it now

oystub commented on src/modules/navigator/mission_rtl.cpp:
```diff
@@ -0,0 +1,199 @@
+
+	} else if (any_winch_commands_backwards && !any_winch_commands_forwards) {
+		should_reverse = false;
+		PX4_INFO("MissionRTl: fast forward because of winch command on the reverse path");
```
Do we want this? Could it be more general to just skip certain types of waypoints? That would make it usable for one-way missions as well.

oystub commented on src/modules/navigator/mission_rtl.cpp:
```diff
@@ -0,0 +1,199 @@
+
+	_mission_handle->on_inactive();  // Tick once to propagate the mission
+	_mission_handle->on_activation();
+}
```
I think we should get the landing location from the mission, and if its close to home, just update the home position?
Mission RTL has a much greater chance of landing at the takeoff location than regular RTL.

oystub replied:
Resolved by doing normal RTL landing

oystub commented on src/modules/navigator/navigator_main.cpp:
```diff
@@ -80,17 +80,19 @@ Navigator::Navigator() :
+	_navigation_mode_array[6] = &_precland;
 #if CONFIG_MODE_NAVIGATOR_VTOL_TAKEOFF
-	_navigation_mode_array[6] = &_vtol_takeoff;
+	_navigation_mode_array[7] = &_vtol_takeoff;
```
Is there a reason to change all of these?

Torjusba replied:
Moved it to 6 now, to change only `vtol_takeoff`. I think it's better to change that one than to potentially leave a hole in the array

oystub commented on src/modules/navigator/rtl_base.h:
```diff
@@ -54,4 +56,43 @@ class RtlBase : public MissionBase
+			if (!success) {
+				n_failed++;
+			}
+		}
```
Are there race conditions here if something else changes the mission? Could there be any realistic reason for this happening?

Torjusba replied:
Yes, if something changes the mission at the same time as MissionRTL activates but I don't think it's very realistic.

We could potentially wrap the entire `on_activation` function inside a retry loop and check the mission id before and after reading, but I'm not sure if it's worth it. Let's discuss in person

oystub commented on src/modules/navigator/rtl_base.h:
```diff
@@ -54,4 +56,43 @@ class RtlBase : public MissionBase
+	*/
+	mission_item_s *getMissionItems()
+	{
+		mission_item_s *arr = (mission_item_s *)calloc(_mission.count, sizeof(mission_item_s));
```
Any reason that they must be 0-initialized?

Torjusba replied:
no longer relevant, fixed it

oystub commented on src/modules/navigator/rtl_base.h:
```diff
@@ -54,4 +56,43 @@ class RtlBase : public MissionBase
+	*/
+	mission_item_s *getMissionItems()
+	{
+		mission_item_s *arr = (mission_item_s *)calloc(_mission.count, sizeof(mission_item_s));
```
Need to check that the allocation is successful, this might be up to `NUM_MISSIONS_SUPPORTED` items.

Could we do some kind of individual item return instead, just calling multiple times to avoid allocating all this memory? I'm a bit concerned that the current method could cause RAM issues.

e.g.
```
bool getMissionItemByIdx(size_t idx, mission_item_s* out)
```

Torjusba replied:
good suggestion, fixed

oystub reviewed (changes requested):
Like the overall concept. Some comments on both details and general functionality.

- I think we should get "land at land point instead of home(takeoff)" implemented, as this greatly reduces risk of bad landings.

- We should think about whether we need to skip some actions (or possibly only whitelist some types of actions).

- Is there a proper guard against triggering M-RTL in MR? If it triggers immediately after MR transition at customer, what will happen if we reverse? Transition?

- Should enforce that M-RTL does not override regular RTL, in case we get an RTL failure first (or the pilot has manually triggered RTL)

oystub commented on Tools/aviant/sitl_docker.sh:
```diff
@@ -0,0 +1,42 @@
+    --network=host \
+    --env=HEADLESS=1 \
+    --env=PX4_SIM_SPEED_FACTOR=3 \
+    --volume="${HOST_PX4_DIR}:${HOST_PX4_DIR}:rw" \
```
is `rw` acutally a valid option for "volume"?

Cannot find it here:
https://docs.docker.com/engine/storage/bind-mounts/

It's basically just the default behavior?

Torjusba replied:
Yes it's default, I was explicit here because it's specified in `Tools/docker_run.sh`

oystub commented on src/modules/commander/px4_custom_mode.h:
```diff
@@ -226,10 +227,14 @@ static inline union px4_custom_mode get_px4_custom_mode(uint8_t nav_state)
 	return custom_mode;
 }
 
 #endif /* DEFINE_GET_PX4_CUSTOM_MODE */
```
Unintended change?

Torjusba committed:
failsafe: remove "Autopilot disengaged" from generic event message

It's not really accurate, since the generic cause can be a lot of
things.
This phrase will likely be more confusing than useful for our pilots.

Torjusba committed:
failsafe: treat RTL flight mode as a failsafe action

In order to get the same override behaviour when the flight mode is
selected manually as with a failsafe.

oystub commented on src/modules/commander/Commander.cpp:
```diff
@@ -45,6 +45,7 @@
 #include "commander_helper.h"
 #include "esc_calibration.h"
 #include "uORB/topics/vehicle_local_position.h"
+#include "uORB/topics/vehicle_status.h"
```
I don't see what change makes this one needed.

oystub commented on src/modules/commander/failsafe/failsafe.h:
```diff
@@ -194,4 +195,3 @@ class Failsafe : public FailsafeBase
 				       );
 
 };
```
Shouldn't need to change this line

oystub commented on src/modules/commander/failsafe/framework.h:
```diff
@@ -284,4 +287,3 @@ class FailsafeBase: public ModuleParams
 				       );
 
 };
-
```
Same here

oystub commented on src/modules/navigator/mission_rtl.cpp:
```diff
@@ -0,0 +1,210 @@
+
+	const mission_s mission_info = _mission_handle->getMissionInfo();
+	float km_backwards{0};
+	float km_forwards{0};
```
I think it would be optimal to use meters for all distances, since this is the norm everywhere else. Then convert to km for the printing if desirable.

oystub commented on src/modules/navigator/mission_rtl.cpp:
```diff
@@ -0,0 +1,210 @@
+		} else {
+			// False by default, since turning around late in the mission
+			// can almost double the flight distance, and we would run
+			// out of battery before reaching home.
```
I think this command is a bit misleading, since it only happens if the lengths are exactly the same, but the comment makes it feel like a more general case.

oystub commented on src/modules/navigator/missionrtl_mission_fast_reverse.cpp:
```diff
@@ -0,0 +1,252 @@
+
+void MissionRtlMissionFastReverse::on_activation(bool from_mission_mode)
+{
+	_home_pos_sub.update();
```
Do we need to handle this one not being set? Probably never happens, home must always be set before takeoff, right?

Any reason that this subscriber isn't initialized before that? Or do uorb subscribers also get the "previous" mesage on the bus before they were created?

Torjusba replied:
Yes it gets the previous message

oystub commented on src/modules/navigator/rtl_base.h:
```diff
@@ -40,6 +40,7 @@
 #pragma once
 
 #include "mission_base.h"
+#include <cstdlib>
```
?

oystub commented on src/modules/commander/failsafe/framework.cpp:
```diff
@@ -546,6 +546,21 @@ void FailsafeBase::getSelectedAction(const State &state, const failsafe_flags_s
+		// Skip MissionRTL if we're falling though, but we need to place
+		// mission RTL here so it can fall through to other actions
+		if (selected_action == Action::MissionRTL &&
+		    state.vehicle_type == vehicle_status_s::VEHICLE_TYPE_FIXED_WING &&
```
It's not super obvious from anywhere in the code (naming, comments, flow) *why* it is only supposed to activate in FW, and the fallback behaviour to normal RTL during landing.

oystub reviewed (approved):
This is looking very good now. I have some suggestions, but none are required.

If the unrelated newlines are required for pre-commit hook to work well, I suggest doing them in a separate commit, to make merging and cherry-picking potentially easier.

I thought a bit about what to call this mode, since it is specifically for fixed-wing only and with many Aviant-specific considerations. Should it be named "aviant(flavour/style/type)" mission rtl to make clear that this is *our* mode?

Do you think we should remove the existing mission rtl (or at least the failsafe enum value) to make it less confusing? Or are we ok with just parameter monitoring?
